BJS2-78048 Интеграция с генератором случайных аватаров юзера

dextomi opened the pull request.

dextomi committed:
1. Реализовал генерацию аватарок через dicebear, поднятый локально
2. Реализовал сохранение аватарок в БД и генерацию ключа, сам ключ, полагаю, будет присваиваться юзеру при регистрации
3. Реализовал генерацию presigned URL по апи для аватарок в S3
4. Сервисный слой для S3 общий, если где-то будут конфликты поправлю, сервисный слой пока не мержили в мастер

dextomi committed:
Поправил конфликты

dextomi committed:
Merge remote-tracking branch 'origin/feature/BJS2-78048-generating-avatars' into feature/BJS2-78048-generating-avatars

# Conflicts:
#	build.gradle.kts
#	src/main/java/school/faang/user_service/service/s3/S3Service.java
#	src/main/resources/application.yaml

dextomi committed:
Поправил некоторые конфликты

dextomi committed:
Поправил некоторые конфликты, удалил тесты для S3Service

dextomi committed:
Надеюсь получилось поправить тесты

dextomi committed:
+ добавил тесты на генерацию пресайнд урл

dextomi committed:
+ добавил тесты на генерацию пресайнд урл

dextomi committed:
поправил свои тесты

Fagan89 commented on src/main/java/school/faang/user_service/client/DiceBearClient.java:
```diff
@@ -0,0 +1,20 @@
+@FeignClient(name = "dicebear-client", url = "${services.dicebear.url}")
+public interface DiceBearClient {
+
+    @GetMapping(value = "/{version}/{style}/{format}", produces = MediaType.IMAGE_PNG_VALUE)
```
я бы Retry добавил бы

Fagan89 commented on src/main/java/school/faang/user_service/service/user/UserProfileService.java:
```diff
@@ -0,0 +1,33 @@
+        UserProfilePic profilePic = user.getUserProfilePic();
+        if (profilePic == null) {
+            throw new FileException("У данного пользователя нет изображения профиля");
+        }
```
в валидатор

Fagan89 commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,63 @@
+                avatarBytes,
+                "avatar",
+                "generated_avatar.png",
+                MediaType.IMAGE_PNG_VALUE
```
через билдер будет покрасивее))

dextomi committed:
Реализовал частично регистрацию для проверки генерации аватарок, сделал метод для генерации аватарок асинхронным

20cognosce commented on src/main/java/school/faang/user_service/client/DiceBearClient.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@FeignClient(name = "dicebear-client",
+        url = "${services.dicebear.url}",
+        configuration = DiceBearClientConfig.class
+)
```
FeignClient подходит только для общения между микросервисами в рамках одного Spring Cloud кластера.
Для запросов во внешние сервисы использовать лучше RestTemplate / WebClient, переделать

20cognosce commented on src/main/java/school/faang/user_service/client/DiceBearClientConfig.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class DiceBearClientConfig {
+
+    @Bean
+    public Logger.Level feignLoggerLevel() {
+        return Logger.Level.FULL;
+    }
+
+    @Bean
+    public Retryer feignRetryer() {
+        return new DiceBearRetryer();
+    }
+
+
+    // Попытка прописать свой RequestInterceptor и обойти дефолтный, как найду решение, вернусь к этому
+    /*@Bean
+    public RequestInterceptor diceBearClientRequestInterceptor() {
+        return template -> {
+            template.removeHeader("Accept");
+        };
+    }
+
+    @Bean
+    public List<RequestInterceptor> diceBearClientRequestInterceptors() {
+        return List.of(diceBearClientRequestInterceptor());
+    }*/
```
Удалить всё это

20cognosce commented on src/main/java/school/faang/user_service/client/DiceBearRetryer.java:
```diff
@@ -0,0 +1,34 @@
+
+@Slf4j
+@NoArgsConstructor
+public class DiceBearRetryer implements Retryer {
```
Удалить

Спринговый @Retryable используй

20cognosce commented on src/main/java/school/faang/user_service/client/DiceBearRetryer.java:
```diff
@@ -0,0 +1,34 @@
+    private int attempt = 0;
+
+    public void continueOrPropagate(RetryableException e) {
+        if (attempt++ >= MAX_ATTEMPTS) {
+            log.error("Превышено максимальное количество попыток для запроса. Ошибка: {}",
+                    e.getMessage());
+            throw e;
+        }
+        try {
+            Thread.sleep(BACKOFF);
+        } catch (InterruptedException ignored) {
+            Thread.currentThread().interrupt();
+            throw e;
+        }
```
ты на это тратил своё время?

20cognosce commented on src/main/java/school/faang/user_service/dto/user/UserProfileDto.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+@AllArgsConstructor
+public class UserProfileDto {
+    private String presignedUrl;
```
Ты аватарку отдаешь или профиль пользователя? Не вводи в заблуждение - сделай тогда контроллер для аватарки, или сделай полноценный профиль со всеми полями

20cognosce commented on src/main/java/school/faang/user_service/controller/user/UserProfileController.java:
```diff
@@ -0,0 +1,31 @@
+    private final UserProfileService profileService;
+    private final UserProfileMapper userProfileMapper;
+
+    @GetMapping("/presigned-url")
+    public ResponseEntity<UserProfileDto> generatePresignedUrl(
```
.

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -60,4 +61,12 @@ services:
\ No newline at end of file
+    bucketName: corpbucket
+    avatarPath: avatar/generated
+    presignedUrl:
+      ttl: 1
```
Почему такой низкий ttl, это секунды? минуты? часы? недели?
лучше в  секундах задавай, чтобы настраивать было проще

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -60,4 +61,12 @@ services:
+    presignedUrl:
+      ttl: 1
+  dicebear:
+    url: http://localhost:3000
```
откуда он у тебя на локалхосте...

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+
+    public User createUser(User user, Long countryId) {
+        UserProfilePic userProfilePic = new UserProfilePic();
+        CompletableFuture<String> avatarUploadFuture = avatarGeneratorService.generateAndUpload();
```
Тебе не нужен CompletableFuture, почему ты не можешь просто идентификатор аватарки получить (ключ) и сразу в юзере сохранить, а генерацию и сохранение в S3 делать в фоне

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+
+        String path;
+        try {
+            path = avatarUploadFuture.get();
```
почему здесь блокируем поток?

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
         return fileDto;
     }
+
+    public User createUser(User user, Long countryId) {
```
И здесь транзакцию тогда навесь

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+            path = avatarUploadFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Произошла ошибка при генерации аватара {}", e.getMessage(), e);
+            path = null;
```
ну что это...

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+
+        String encodedPassword = passwordEncoder.encode(user.getPassword());
+
+        String path;
```
это не path, это ключ в нашем s3 (идентификатор ресурса)

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+        userProfilePic.setSmallFileId(path);
+        user.setUserProfilePic(userProfilePic);
+        user.setCountry(country);
+        user.setPassword(encodedPassword);
```
а нафиг он тебе нужен, если ты его нигде не проверяешь, просто так зависимость на Spring Security притянул в проект
если делаешь фичу - делай до конца, ну... или не делай тогда

20cognosce replied:
Если не понятно как делать - спрашивай, почему ты на встречах не рассказывал ничего что ты тут пароли хэшировать собрался и Security на всех эндпоинтах прикручивать? 

Не удивительно, что ты с задачей средней сложности сидел месяц - потому что половину того, за что ты тут брался, делать было не надо вовсе

20cognosce commented on src/main/java/school/faang/user_service/config/security/SecurityConfig.java:
```diff
@@ -0,0 +1,33 @@
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) {
+        try {
+            http
+                    .csrf(AbstractHttpConfigurer::disable)
+                    .authorizeHttpRequests(authz -> authz
+                            .anyRequest().permitAll()
+                    );
+
+            return http.build();
```
ну вот зачем нам это в этой задаче

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserProfileService.java:
```diff
@@ -0,0 +1,32 @@
+    private final S3Service s3Service;
+    private final UserValidation userValidation;
+
+    public String generatePresignedUrl(long userId, boolean isSmall) {
```
ты находишься в UserProfileService 
думай о всех методах, которые ты тут пишешь в контексте UserProfileService 

если мы работаем с профилем юзера, откуда здесь появляется PresignedUrl? PresignedUrl - это домен S3, а в домене UserProfile только аватарки пользователя, здесь мы должны в контексте аватарок весь код писать

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -108,4 +119,33 @@ private S3FileDto downloadFile(Function<UserProfilePic, String> fileIdExtractor,
+                    return new EntityNotFoundException(String.format(USER_COUNTRY_NOT_FOUND, countryId));
+                });
+
+        String encodedPassword = passwordEncoder.encode(user.getPassword());
```
и почему ты пароль достаешь из юзера, передавай отдельным параметром

20cognosce commented on src/main/java/school/faang/user_service/service/user/UserServiceFacade.java:
```diff
@@ -33,14 +36,23 @@ public UserProfilePic uploadAvatar(MultipartFile file) {
+    public UserRegisterResponseDto registerUser(UserRegisterRequestDto userRegisterRequestDto) {
+        User user = userMapper.toUserEntity(userRegisterRequestDto);
+        user.setPassword(userRegisterRequestDto.getPassword());
+        User savedUser = userService.createUser(user, userRegisterRequestDto.getCountryId());
```
наверно проще просто DTO в метод аргументом передать в этом случае

20cognosce commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Async("avatar-generator-executor")
+    public CompletableFuture<String> generateAndUpload() {
+        userContext.setUserId(0); // Явно прописал 0, т.к не знаю, как убрать необходимость в хедере, ничего не помогло
```
а кто у тебя его требует? ты в сторонний сервис ведь тут идешь, ему зачем x-user-id?

20cognosce commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,70 @@
+
+        try {
+            MultipartFile avatarFile = generateAvatar();
+            log.info("Аватар успешно сгенерирован");
```
если ты будешь разбирать аварию по логам то в этой строчке ты захочешь ругаться, потому что вроде лог есть, но ни черта по нему понять не возможно и никак это в анализе использовать не получится

какой размер, с какими параметрами

20cognosce commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,70 @@
+        return CustomMultipartFile.builder()
+                .content(avatarBytes)
+                .name("avatar")
+                .originalFilename("generated_avatar.png")
```
у всех юзеров одинаковые?

20cognosce commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,70 @@
+            MultipartFile avatarFile = generateAvatar();
+            log.info("Аватар успешно сгенерирован");
+
+            return CompletableFuture.completedFuture(s3Service.uploadFile(avatarFolder, avatarFile));
```
не используй CompletableFuture, используй ExecutorService

20cognosce commented on src/main/java/school/faang/user_service/service/avatar/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,70 @@
+    @Value("${services.s3.avatarPath}")
+    private String avatarFolder;
+
+    @Async("avatar-generator-executor")
```
и если уже тут асинк, сколько нам нужно асинков? пять ли, десять
не понятно)))

20cognosce reviewed (changes requested):
пожалуйста обработай все комментарии и подготовь пул доскональных вопросов к дейлику, чтобы мы друг друга поняли и ты без проблем доделал задачу

dextomi committed:
1. Удалил все связанное с presignedUrl за ненадобностью;
2. Убрал security и удалил зависимости;
3. Сделал генерацию аватарок синхронной;
4. Переделал FeignClient на RestTemplate;
5. Перенес логи по генерации в регистрацию, т.к только там можно получить информацию по юзеру
6. Оставил метод по генерации ключа и сохранению аватарки в S3 как есть, т.к имеющийся метод не принимает ключ в качестве параметра, а редактировать метод или создавать новый через перегрузку как будто бы не нужно;
7. Теперь createUser принимает дто вместо сущности

20cognosce commented on src/main/java/school/faang/user_service/client/DiceBearRestTemplate.java:
```diff
@@ -0,0 +1,30 @@
+    private String baseUrl;
+
+    public byte[] getAvatar(String version, String style, String format, String seed) {
+        String url = String.format("%s/%s/%s/%s?seed=%s", baseUrl, version, style, format, seed);
```
UriComponentsBuilder

20cognosce commented on src/main/java/school/faang/user_service/config/http/RestTemplateConfig.java:
```diff
@@ -0,0 +1,14 @@
+@Configuration
+public class RestTemplateConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
```
чей он
